client/wsclient.go

```diff
@@ -80,13 +104,53 @@ func (c *wsClient) Start(ctx context.Context, settings types.StartSettings) erro
 }
 
 func (c *wsClient) Stop(ctx context.Context) error {
+	c.isStopped.Store(true)
+
 	// Close connection if any.
 	c.connMutex.RLock()
 	conn := c.conn
 	c.connMutex.RUnlock()
 
 	if conn != nil {
-		_ = conn.Close()
+		// Shut down the sender and any other background processors.
+		c.stopBGProcessors <- struct{}{}
+		select {
+		case <-c.bgProcessorsStopped:
+		case <-ctx.Done():
+			_ = c.conn.Close()
+			return errEarlyStop
+		}
+
+		defaultCloseHandler := conn.CloseHandler()
+		closed := make(chan struct{})
+
+		// The server should respond with a close message of its own, which will
+		// trigger this callback. At this point the close sequence has been
+		// completed and the TCP connection can be gracefully closed.
+		conn.SetCloseHandler(func(code int, text string) error {
```

Comment:
Same concurrency question for `SetCloseHandler`.

Reply:
`SetCloseHandler` only sets the close handler function, which will be called later when the client receives a close message from the server. We know that all writes have stopped once `bgProcessingStopped` has fired, which we wait on above in this method.